### DISCUSSION TITLE
Liquidsoap 2.2.5 build for Ubuntu 24.04.1 LTS?

### DISCUSSION THREAD

**Discussion by NicoxDJ** [savonet/liquidsoap#4127] [Installation] [answered]
### Description

I tried the debian package, but i can't install it because some packages are missing:
```
E: Unable to locate package libavcodec59
E: Unable to locate package libavdevice59
E: Unable to locate package libavfilter8
E: Unable to locate package libavformat59
E: Unable to locate package libavutil57
E: Unable to locate package libswresample4
E: Unable to locate package libswscale6
```
There are newer ones for 24.04.1
I couldin't install them on 22.04 neither, because there are *58

**Comment by vitoyucepi** [comment]
Hi @NicoxDJ,
It seems you're trying to use the wrong packages.
For ubuntu:24.04 we have the following options
- use containers;
- use 2.2.4 from ubuntu repos;
- install from opam;
- build from source;
- use release artifacts for the 2.3.x-rolling version.

For ubuntu:22.04 we have the following options
- use containers;
- use release artifacts for any version from 2.1.1 to 2.2.5;
- use 2.0.2 from ubuntu repos;
- install from opam;
- build from source.

You're not supposed to use release artifacts for Debian on Ubuntu because the versions of the dependencies differ.
For example, for ubuntu jammy x86-64 you should use the release artifact available at https://github.com/savonet/liquidsoap/releases/download/v2.2.5/liquidsoap_2.2.5-ubuntu-jammy-1_amd64.deb.

The installation manual is available at https://www.liquidsoap.info/doc-2.2.5/install.html. It covers most of the advised methods.
I prefer to use container images, which are available at https://hub.docker.com/r/savonet/liquidsoap.
The building from opam is also a viable solution https://www.liquidsoap.info/doc-2.2.5/install.html#install-using-opam.

**Reply by NicoxDJ** [reply]
Well, right now i'm runing 2.2.4 from the Ubuntu distro, but i need to use AutoCue which requires 2.2.5, because it's crashing with error:
```
At /var/azuracast/autocue/autocue.liq, line 59, char 0-34:
let settings.autocue.cue_file = ()

Error 4: Undefined variable settings
```

I tried and I can't build 2.3.x by myself:
```
File "src/lang/queues.ml", line 28, characters 28-31:
28 |       try flush_elements_f (pop q :: elements) with Empty -> List.rev elements
                                 ^^^
Error: Unbound value pop
make: *** [Makefile:4: build] Error 1 
```

I can't install 2.2.5 with opam neither:
```
<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
⬇ retrieved liquidsoap.2.2.5  (cached)
⬇ retrieved liquidsoap-core.2.2.5  (cached)
[ERROR] The compilation of liquidsoap-core.2.2.5 failed at "dune build -p liquidsoap-core -j 7 --promote-install-files=false @install".

#=== ERROR while compiling liquidsoap-core.2.2.5 ==============================#
# context     2.1.5 | linux/x86_64 | ocaml.4.14.1 | https://opam.ocaml.org#abac1deff0ec6249cf93f6b25ffdc49ae1caf04c
# path        ~/.opam/default/.opam-switch/build/liquidsoap-core.2.2.5
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p liquidsoap-core -j 7 --promote-install-files=false @install
# exit-code   1
# env-file    ~/.opam/log/liquidsoap-core-82362-5f5d54.env
# output-file ~/.opam/log/liquidsoap-core-82362-5f5d54.out
### output ###
# [...]
# (cd _build/default && /usr/bin/ocamlc.opt -w -40 -g -bin-annot -I src/core/.liquidsoap_ffmpeg.objs/byte -I /home/nicox/.opam/default/lib/angstrom -I /home/nicox/.opam/default/lib/backoff -I /home/nicox/.opam/default/lib/bigstringaf -I /home/nicox/.opam/default/lib/bytes -I /home/nicox/.opam/default/lib/camlp-streams -I /home/nicox/.opam/default/lib/camomile/lib -I /home/nicox/.opam/default/li[...]
# File "src/core/encoder/encoders/ffmpeg_encoder_common.ml", line 118, characters 54-60:
# 118 |     | `String layout -> `Int64 Avutil.Channel_layout.(get_id (find layout))
#                                                             ^^^^^^
# Error: Unbound value get_id
# (cd _build/default && /usr/bin/ocamlc.opt -w -40 -g -bin-annot -I src/core/.liquidsoap_ffmpeg.objs/byte -I /home/nicox/.opam/default/lib/angstrom -I /home/nicox/.opam/default/lib/backoff -I /home/nicox/.opam/default/lib/bigstringaf -I /home/nicox/.opam/default/lib/bytes -I /home/nicox/.opam/default/lib/camlp-streams -I /home/nicox/.opam/default/lib/camomile/lib -I /home/nicox/.opam/default/li[...]
# File "src/core/stream/ffmpeg_raw_content.ml", line 83, characters 54-58:
# 83 |             (Channel_layout.get_description ?channels:None)
#                                                            ^^^^
# Error: The function applied to this argument has type
#          Avutil.Channel_layout.t -> string
# This argument cannot be applied with label ?channels



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build liquidsoap-core 2.2.5
└─ 
```

containers is not an option too.

Can you help to fix at least one of these 3 errors?

**Reply by vitoyucepi** [reply, marked as answer]
So you're using ubuntu:24.04 but don't want to use containers and want to use autocue.
May I ask you to share your installation script?

Here's a script I used on a clean VM
```sh
OPAM_VERSION=2.2.1;
OCAML_VERSION=4.14.2;
```
```sh
sudo apt-get update;
sudo apt-get dist-upgrade;
sudo apt-get install --no-install-recommends -y curl ca-certificates patch unzip bubblewrap git bzip2 gcc make libc6-dev g++;
sudo curl -sS -L -o /usr/local/bin/opam "https://github.com/ocaml/opam/releases/download/$OPAM_VERSION/opam-$OPAM_VERSION-x86_64-linux";
sudo chmod +x /usr/local/bin/opam;
opam init --bare -n;
opam switch create "$OCAML_VERSION";
eval $(opam env --switch="$OCAML_VERSION");
opam install --confirm-level=unsafe-yes liquidsoap ffmpeg;
```

**Reply by NicoxDJ** [reply]
After hitting this step: `opam switch create "4.14.2"`

This is what i've got:
```
<><> Installing new switch packages <><><><><><><><><><><><><><><><><><><><><><>
Switch invariant: ["ocaml-base-compiler" {= "4.14.2"} | "ocaml-system" {= "4.14.2"}]

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
⬇ retrieved ocaml-config.2  (cached)
∗ installed base-bigarray.base
∗ installed base-threads.base
∗ installed base-unix.base
∗ installed host-arch-x86_64.1
∗ installed host-system-other.1
∗ installed ocaml-options-vanilla.1
⬇ retrieved ocaml-base-compiler.4.14.2  (cached)
[ERROR] The compilation of ocaml-base-compiler.4.14.2 failed at "./configure --prefix=/home/nicox/.opam/4.14.2 --docdir=/home/nicox/.opam/4.14.2/doc/ocaml
        -C".

#=== ERROR while compiling ocaml-base-compiler.4.14.2 =========================#
# context     2.2.1 | linux/x86_64 |  | https://opam.ocaml.org#abac1deff0ec6249cf93f6b25ffdc49ae1caf04c
# path        ~/.opam/4.14.2/.opam-switch/build/ocaml-base-compiler.4.14.2
# command     ~/.opam/opam-init/hooks/sandbox.sh build ./configure --prefix=/home/nicox/.opam/4.14.2 --docdir=/home/nicox/.opam/4.14.2/doc/ocaml -C
# exit-code   1
# env-file    ~/.opam/log/ocaml-base-compiler-12514-d9a409.env
# output-file ~/.opam/log/ocaml-base-compiler-12514-d9a409.out
### output ###
# bwrap: loopback: Failed RTM_NEWADDR: Operation not permitted



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build ocaml-base-compiler 4.14.2
└─ 
┌─ The following changes have been performed (the rest was aborted)
│ ∗ install base-bigarray         base
│ ∗ install base-threads          base
│ ∗ install base-unix             base
│ ∗ install host-arch-x86_64      1
│ ∗ install host-system-other     1
│ ∗ install ocaml-options-vanilla 1
└─ 
# Run eval $(opam env --switch=4.14.2) to update the current shell environment
Switch initialisation failed: clean up? ('n' will leave the switch partially installed) [y/n] y
```

**Reply by NicoxDJ** [reply]
I've manage to install it, after fixing this bwrap with [this](https://github.com/ocaml/opam/issues/5968#issuecomment-2151748424)

Now i successfully installed 2.2.5-dev, but i've got another error:
```
At /var/azuracast/stations/station/config/liquidsoap.liq, line 43, char 18-32:
https_transport = http.transport.ssl()

Error 5: this value has no method `ssl`
  Its type is {unix : http_transport.{default_port : int, name : string, protocol : string}
}.
```

And if i commented it i've got:
```
At /var/azuracast/stations/station/config/liquidsoap.liq, line 454, char 8-14:
radio = ladspa.master_me(

Error 4: Undefined variable ladspa
```

and if i comment it i've got:
```
At /var/azuracast/stations/station/config/liquidsoap.liq, line 431, char 4-52:
    %mp3(samplerate=44100, stereo=true, bitrate=128),

Error 8: Unsupported format: %mp3(samplerate=44100, stereo=true, bitrate=128)
```
Why just nothing works :(

**Reply by vitoyucepi** [reply]
This will probably help https://github.com/ocaml/opam/issues/5968, but may cause bugs later.
The update with the changes has been rolled up, then rolled back, and the current status of the new version should be tracked at https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/2064672.

I think you can try
```sh
sudo curl -sS -L \
  -o /etc/apparmor.d/bwrap-userns-restrict \
  https://gitlab.com/apparmor/apparmor/-/blob/37cac653d115b9d16afe89188b6872af55dd75b8/profiles/apparmor/profiles/extras/bwrap-userns-restrict
```

**Reply by vitoyucepi** [reply]
> this value has no method `ssl`

Built without `ssl`.
- `opam install ssl`.
- `opam install tls` for alternative library, but you should also change your script.

> Undefined variable `ladspa`

Built without `ladspa`. `opam install ladspa`.

> Unsupported format: `%mp3`

Built without mp3 encoder.
- Encode using `%ffmpeg`.
- `opam install lame`.

**Reply by NicoxDJ** [reply]
Thank you! I installed them and so far everything looks to be ok.
This is the command that i used:
`opam install --confirm-level=unsafe-yes liquidsoap ffmpeg taglib mad lame vorbis cry samplerate ssl ladspa;`
I want to add WAV and Flac support to the list too.
Where can i find the complete list with plugins that Liquidsoap uses? I want to install them all, so to be the most compatible one?

**Reply by vitoyucepi** [reply]
The `liquidsoap --build-config` command will show you all the suggested opam packages.
But do you need to install them all?

**Reply by mailinglists35** [reply]
> For ubuntu:22.04 we have the following options
> 
> * use containers;
> * use release artifacts for any version from 2.1.1 to 2.2.5;
> * use 2.0.2 from ubuntu repos;
> * install from opam;
> * build from source.
> 

this means no hope for a release artifacts of 2.3.x for ubuntu jammy? :-(

**Reply by vitoyucepi** [reply]
Hi @mailinglists35,

> no hope for a release artifacts of 2.3.x for ubuntu jammy?

I'm not aware of any plans to continue supporting ubuntu:22.04, at least in the official liquidsoap github repository.
If I were you, I would consider upgrading to ubuntu:24.04, using container images, or building from opam. As a last resort, I'd probably try to hire someone to do the server maintenance for me. But I've never hired someone for a no-code solution of this kind before.
